File: DevExpress-Demo/CS/NavigationDemo.Wpf/ViewModels/ReportLibraryViewModel.cs
using System;
using System.Windows;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using DevExpress.DemoData.Helpers;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Xpf.Core;

namespace NavigationDemo {
    public class ReportLibraryViewModel {
        protected virtual IFolderBrowserDialogService FolderBrowserDialogService { get { return this.GetService<IFolderBrowserDialogService>(); } }

        public IEnumerable<ReportLibraryNode> Nodes { get; private set; }
        public virtual ReportLibraryNode SelectedNode { get; set; }

        public ObservableCollection<ReportLibraryNode> CheckedReportNodes { get; set; }

        public ReportLibraryViewModel() {
            var reportsFolder = DataFilesHelper.FindFile("Reports", DataFilesHelper.DataPath);
            Nodes = new ObservableCollection<ReportLibraryNode>(GetFolderChildren(reportsFolder));
            CheckedReportNodes = new ObservableCollection<ReportLibraryNode>();
            SelectedNode = Nodes.First().Children.FirstOrDefault(x => x.Document != null);
        }

        public void OnNodePositionChanging(DragRecordOverEventArgs e) {
            var isTargetFolder = ((ReportLibraryNode)e.TargetRecord).IsFolder;
            var isInside = e.DropPosition == DropPosition.Inside;
            if((!isTargetFolder && isInside)
                || (isTargetFolder && !isInside && !SelectedNode.IsFolder)) 
                e.Effects = DragDropEffects.None;
             else 
                e.Effects = DragDropEffects.All;
        }
        public void OnCheckStateChanged(ReportLibraryNode node, object isChecked) {
            if((bool?)isChecked == true && !node.IsFolder && !CheckedReportNodes.Contains(node))
                CheckedReportNodes.Add(node);
            else
                CheckedReportNodes.Remove(node);
        }
        public virtual void SaveReports() {
            if(FolderBrowserDialogService.ShowDialog())
                SaveReportsCore(FolderBrowserDialogService.ResultPath);            
        }
        public virtual bool CanSaveReports() {
            return CheckedReportNodes.Any();
        }
        void SaveReportsCore(string targetFolder) {
            foreach (var node in CheckedReportNodes) {
                var fileName = Path.GetFileName(node.Document);
                var targetPath = Path.Combine(targetFolder, fileName);
                try {
                    File.Copy(node.Document, targetPath, overwrite: true);
                } catch { }               
            }
        }
        static ReportLibraryNode CreateFolderNode(string path) {
            var node = ReportLibraryNode.Create(Path.GetFileName(path));
            foreach(var child in GetFolderChildren(path))
                node.Children.Add(child);
            return node;
        }
        static IEnumerable<ReportLibraryNode> GetFolderChildren(string path) {
            return Directory.GetDirectories(path).Select(x => CreateFolderNode(Path.Combine(path, x)))
                .Concat(Directory.GetFiles(path).Select(x => ReportLibraryNode.Create(Path.GetFileNameWithoutExtension(x), x)));
        }
    }
    public class ReportLibraryNode {
        public static ReportLibraryNode Create(string name, string document = null) {
            var node = ViewModelSource.Create(() => new ReportLibraryNode());
            node.Name = name;
            node.Document = document;
            return node;
        }
        public virtual string Name { get; set; }
        public virtual string Document { get; set; }
        readonly ObservableCollection<ReportLibraryNode> children;
        public ObservableCollection<ReportLibraryNode> Children { get { return children; } }
        public virtual bool IsFolder { get { return string.IsNullOrEmpty(Document); } }
        protected ReportLibraryNode() {
            children = new ObservableCollection<ReportLibraryNode>();
        }
        public override string ToString() {
            return Name;
        }
    }
}

File: DevExpress-Demo/VB/NavigationDemo.Wpf/ViewModels/ReportLibraryViewModel.vb
Imports System
Imports System.Windows
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.IO
Imports System.Linq
Imports DevExpress.DemoData.Helpers
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpf.Core

Namespace NavigationDemo

    Public Class ReportLibraryViewModel

        Private _Nodes As IEnumerable(Of NavigationDemo.ReportLibraryNode)

        Protected Overridable ReadOnly Property FolderBrowserDialogService As IFolderBrowserDialogService
            Get
                Return GetService(Of IFolderBrowserDialogService)()
            End Get
        End Property

        Public Property Nodes As IEnumerable(Of ReportLibraryNode)
            Get
                Return _Nodes
            End Get

            Private Set(ByVal value As IEnumerable(Of ReportLibraryNode))
                _Nodes = value
            End Set
        End Property

        Public Overridable Property SelectedNode As ReportLibraryNode

        Public Property CheckedReportNodes As ObservableCollection(Of ReportLibraryNode)

        Public Sub New()
            Dim reportsFolder = DataFilesHelper.FindFile("Reports", DataFilesHelper.DataPath)
            Nodes = New ObservableCollection(Of ReportLibraryNode)(GetFolderChildren(reportsFolder))
            CheckedReportNodes = New ObservableCollection(Of ReportLibraryNode)()
            SelectedNode = Enumerable.First(Nodes).Children.FirstOrDefault(Function(x) Not Equals(x.Document, Nothing))
        End Sub

        Public Sub OnNodePositionChanging(ByVal e As DragRecordOverEventArgs)
            Dim isTargetFolder = CType(e.TargetRecord, ReportLibraryNode).IsFolder
            Dim isInside = e.DropPosition = DropPosition.Inside
            If Not isTargetFolder AndAlso isInside OrElse isTargetFolder AndAlso Not isInside AndAlso Not SelectedNode.IsFolder Then
                e.Effects = DragDropEffects.None
            Else
                e.Effects = DragDropEffects.All
            End If
        End Sub

        Public Sub OnCheckStateChanged(ByVal node As ReportLibraryNode, ByVal isChecked As Object)
            If CType(isChecked, Boolean?) = True AndAlso Not node.IsFolder AndAlso Not CheckedReportNodes.Contains(node) Then
                CheckedReportNodes.Add(node)
            Else
                CheckedReportNodes.Remove(node)
            End If
        End Sub

        Public Overridable Sub SaveReports()
            If FolderBrowserDialogService.ShowDialog() Then SaveReportsCore(FolderBrowserDialogService.ResultPath)
        End Sub

        Public Overridable Function CanSaveReports() As Boolean
            Return CheckedReportNodes.Any()
        End Function

        Private Sub SaveReportsCore(ByVal targetFolder As String)
            For Each node In CheckedReportNodes
                Dim fileName = Path.GetFileName(node.Document)
                Dim targetPath = Path.Combine(targetFolder, fileName)
                Try
                    File.Copy(node.Document, targetPath, overwrite:=True)
                Catch
                End Try
            Next
        End Sub

        Private Shared Function CreateFolderNode(ByVal path As String) As ReportLibraryNode
            Dim node = ReportLibraryNode.Create(IO.Path.GetFileName(path))
            For Each child In GetFolderChildren(path)
                node.Children.Add(child)
            Next

            Return node
        End Function

        Private Shared Function GetFolderChildren(ByVal path As String) As IEnumerable(Of ReportLibraryNode)
            Return Directory.GetDirectories(path).Select(Function(x) CreateFolderNode(IO.Path.Combine(path, x))).Concat(Directory.GetFiles(path).Select(Function(x) ReportLibraryNode.Create(IO.Path.GetFileNameWithoutExtension(x), x)))
        End Function
    End Class

    Public Class ReportLibraryNode

        Public Shared Function Create(ByVal name As String, ByVal Optional document As String = Nothing) As ReportLibraryNode
            Dim node = ViewModelSource.Create(Function() New ReportLibraryNode())
            node.Name = name
            node.Document = document
            Return node
        End Function

        Public Overridable Property Name As String

        Public Overridable Property Document As String

        Private ReadOnly childrenField As ObservableCollection(Of ReportLibraryNode)

        Public ReadOnly Property Children As ObservableCollection(Of ReportLibraryNode)
            Get
                Return childrenField
            End Get
        End Property

        Public Overridable ReadOnly Property IsFolder As Boolean
            Get
                Return String.IsNullOrEmpty(Document)
            End Get
        End Property

        Protected Sub New()
            childrenField = New ObservableCollection(Of ReportLibraryNode)()
        End Sub

        Public Overrides Function ToString() As String
            Return Name
        End Function
    End Class
End Namespace
